let the influxdb API set a database

ryantxu opened the pull request:
Here is a simple option for #7990.  Unlike #7948, this is only the API and is not exposed generally in the UI.

My real motivation for this is to support:
https://github.com/NatelEnergy/grafana-influx-admin

ryantxu committed:
allow setting the database

torkelo commented on public/app/plugins/datasource/influxdb/datasource.ts:
```diff
@@ -157,7 +157,7 @@ export default class InfluxDatasource {
   };
 
-  metricFindQuery(query) {
+  metricFindQuery(query: string, db?: string) {
```
the second param is already taken, not sure it's used yet but we plan to. Needs to be an object (for query options). For example, it's where Grafana can add time range limits to this function and make other specific data source option (if we have specific UI for template variables for example). 

So please change this to an optional object param named options. And on that have a property named database.

ryantxu replied:
done.  Holler if you you want any other changes

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into influx-db-query2

ryantxu committed:
pass database parameter in the options

ryantxu commented:
@torkelo anthing else you want to see with this?

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into influx-db-query2

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into influx-db-query2

ryantxu committed:
pass the options along with a _seriesQuery

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into influx-db-query2

ryantxu commented:
Any chance this will make it into 4.3?  Anything else you would like me to change/improve?  thanks!

ryantxu committed:
use targets[0] as the options

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into influx-db-query2

ryantxu committed:
Merge branch 'influx-db-query2' of https://github.com/ryantxu/grafana into influx-db-query2

torkelo commented on public/app/plugins/datasource/influxdb/datasource.ts:
```diff
@@ -85,7 +85,7 @@ export default class InfluxDatasource {
     allQueries = this.templateSrv.replace(allQueries, scopedVars);
 
-    return this._seriesQuery(allQueries).then((data): any => {
+    return this._seriesQuery(allQueries, targets[0]).then((data): any => {
```
this seems like it will cause issues, using the first query only for the options.

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into influx-db-query2

ryantxu committed:
use the original options parameter

ryantxu commented:
@torkelo good point.  I changed it use the options from the container not each query.

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into influx-db-query2

ryantxu committed:
merge with master

ryantxu commented:
I updated this so that is merges with master again... anything else I should do @torkelo?

ryantxu committed:
merge with master

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into influx-db-query2

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into influx-db-query2

* grafana/master: (248 commits)
  add AWS/NATGateway metrics (#9202)
  docs: minor docs update
  docs: added v4.4 to docs archive
  docs: uppdated getting started with images and text, added info on query inspector to version45 and trouble shooting (#9193)
  reduce readme to be about the code
  adds note about closing #8876
  build: fixed broken build
  Optimize some wrong usage and spelling
  show only label name in label matcher (#9167)
  (prometheus) cache metric suggest query result (#9182)
  Have 'Hide series with only zeros' ignore nulls (#9179)
  docs: updated image in docs (#9184)
  fix: graph right y label is now correctly possitioned, fixes #9172
  Fix sortByName and percentileOfSeries in gfunc.js (#9169)
  docs: updated docs with v4.5 beta links
  Update CHANGELOG.md
  version bump to v4.5-beta1
  fix: fixed jshint issue
  fix: Elasticsearch and adhoc filters fix, fixes #9165
  docs: update to http api admin section
  ...

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into influx-db-query2

* grafana/master: (35 commits)
  Update CHANGELOG.md
  Update CHANGELOG.md
  fixed some images (#9237)
  release: bumped version to v4.5.0 stable
  docs: minor update
  plugins: add styles for ol tags in markdown
  docs: minor fixes to images
  Docs image updates (#9225)
  fix: improve behavior when switching back and forth between x-axis modes, fixes #9229
  fixes for metrics tab when data source was not found
  provide ace editor for external datasource plugin (#9224)
  ux: increased code editor auto complete width from 320px to 550px, fixes #9203
  docs: windows - add note about ini comments
  prometheus: added completer unit test, #9208
  docs: minor update
  docs: kiosk mode options add to playlist doc
  influxdb: small css fix for order by in query editor
  style: corrected indentation in sass file
  replaced old images and gifs with new ones (#9217)
  ux: success/error alerts refactoring, #9214
  ...

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into influx-db-query2

* grafana/master: (112 commits)
  changelog: adds note about closing #5873
  Docs new updates (#9324)
  Update CHANGELOG.md
  Update latest.json
  cleanup: removed unused file
  tech: remove bower and moved remaining bower dependencies to npm
  tech: cleanup and fixed build issue
  tech: upgraded angularjs and moved dependency from bower to npm, closes #9327
  follow go idiom and return error as second param
  tech: updated tsconfig
  docker: adds alertmanager to prometheus fig
  tech: more tslint rules
  tech: removing unused variables from typescript files, and making tslint rules more strict
  context is reserved for go's context
  make ds a param for Query
  remove batch abstraction
  rename executor into tsdbqueryendpoint
  remove unused structs
  refactor response flow
  tech: removed test component
  ...

ryantxu committed:
fix merge issue

ryantxu commented:
@torkelo anything I can do to bump this issue?  Will be great to get a reasonable influx admin now that that is completely removed from their distribution.

torkelo commented:
will try to find time to review this, sorry for the time it has taken :(

torkelo commented:
Wont the requests be stoped by the validation in the data proxy code that validates the database is the same as the one set in datasource settings ? https://github.com/grafana/grafana/blob/master/pkg/api/pluginproxy/ds_proxy.go#L168

ryantxu commented:
Good point -- i have only tested with the 'direct' connection, but we should make sure it works with both.  It looks like this would require a change in the influx settings and/or proxy.  I started down this path in #7948, but that was decided to be too specific.

Do you have opinions on how to approach this?  Is this too rare to be generally useful?  

thanks

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into influx-db-query2

* grafana/master: (282 commits)
  Update CHANGELOG.md
  Fix heatmap Y axis rendering (#9580)
  prometheus: add builtin template variable as range vectors
  Note for #5457
  fix: fixed prometheus step issue that caused browser crash, fixes #9575
  changelog: adds note about closing #9551
  fix: getting started panel and mark adding data source as done, fixes #9568
  pluginloader: esModule true for systemjs config
  Fixes for annotations API (#9577)
  When Messasge field is set for an alert, map it to the output field in a Sensu check result. If Message is empty, send "Grafana Metric Condition Met"
  fix vector range
  allow ":" character for metric name
  build: added imports of rxjs utility functions
  fix template variable expanding
  Update CHANGELOG.md
  annotations: quote reserved fields (#9550)
  Update CHANGELOG.md
  Update CHANGELOG.md
  Update CHANGELOG.md
  fix: fixed color pickers that were broken in minified builds, fixes #9549
  ...

ryantxu committed:
Add a setting to allow DB queries

ryantxu committed:
run go fmt

bergquist commented:
First.. Thank you for really taking the time to contribute to Grafana. Both the admin app the this PR.

I'm not sure it makes sense to have this in the Core Influxdb datasource in Grafana.

The purpose of the metrics client in Grafana is only querying graph data.
The admin app have a much broader use case. So I think it makes more sense to create an admin specific data source in the app rather than reusing the core influxdb database. That would reduce the requirements for the core datasource and the datasource in your app could be developed at a pace thats suitable for the admin app. 

Or maybe I'm missing a good reason for using the same datasource?

ryantxu committed:
merge with master

ryantxu committed:
fix lint problems

ryantxu committed:
fix lint problems

ryantxu committed:
fix lint problems

ryantxu committed:
fix lint problems

ryantxu commented:
@daniellee I just got this to merge cleanly with master.  

Looking at it again, I vote that we:
1. remove the `allowDatabaseQuery` option
1. entirely remove the check in [ds_proxy.go](https://github.com/ryantxu/grafana/blob/influx-db-query2/pkg/api/pluginproxy/ds_proxy.go#L181) that makes sure you don't change the database

This will simplify this change dramatically, and I don't think the check adds any real security because:
* if the credentials don't allow access to the DB, it won't have access
* the influx query syntax lets you change databases using the raw query, not just the parameter.  For example you can do:

```
?db=_internal&q=SELECT numSeries FROM "database" LIMIT 10
```
same results as
```
?db=something&q=SELECT numSeries FROM "_internal".."database" LIMIT 10
```

torkelo commented:
Interesting that InfluxDB has no security checks at all for the db parameter. Makes it very missleading , and also the option in Grafana becomes a bit missleading as I know a lot use multiple influxdb data sources (in seperate orgs) to handle data isolation, wonder if everyone knows that InfluxDB has no security checks for this param.

ryantxu commented:
> InfluxDB has no security checks at all for the db parameter

InfluxDB *does* check if you have access to the db based on your credentials.  My point is that the query syntax lets you switch between databases using the *q* parameter, not just the *db* parameter.

Using a grafana datasource with credentials that can access other databases means the end user can request data from other databases.  It is the credentials that define the data isolation.  This is very similar to SQL.

I think the added check in grafana is misleading since the security constraints must be set in influxdb if you really want data security.  

Maybe we should add a note next to the DB selection in the UI?

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into influx-db-query2

* grafana/master:
  docker: change port for prom random data scrape target

ryantxu committed:
fix merge error

ryantxu commented:
If we remove the db parameter check in the proxy, the change could look somethig like this:
https://github.com/grafana/grafana/compare/master...ryantxu:influx-db-query3

We may also want to add a note next to the database selector that says something like:
```
Note: The influx query syntax allows changing databases in the query.  To support database isolation, make sure the influx credentials limit view access.
```

torkelo commented:
yes, we definitely need that security notice

ryantxu committed:
remove

ryantxu committed:
allow any database for influx proxy

ryantxu commented:
I just updated the PR to remove allowDatabaseQuery and the proxy check since it does not really prevent access to other databases.

Here is the note I added:

![image](https://user-images.githubusercontent.com/705951/37461962-844e53b8-2850-11e8-953b-652d0fa9568d.png)

I'm sure the wording can be improved :)

ryantxu commented:
We may want the grafana proxy to disable POST requests to influx, however right now GET or POST can both be used for operations that modify the data.  (GET requests have a note that it will be disabled in the future).

In any case, people should always configure security/access in InfluxDB and not assume that grafana's access is limited to the configured database and read-only.

ryantxu commented:
@daniellee let me know if you need anything from me on this

torkelo commented:
This looks good to me.

ryantxu commented:
holy smokes -- it really was 1 year and 1 day.  Thanks
